lonestar/analytics/cpu/random-walk/RandomWalk.cpp

```diff
@@ -0,0 +1,631 @@
+
+              walk.push_back(std::move(nbr));
+              types_vec.push_back(type);
+
+            } else {
+              uint32_t curr = walk[walk.size() - 1];
+              uint32_t prev = walk[walk.size() - 2];
+
+              uint32_t p1 = types_vec.back();  //last element of types_vec
+
+              //acceptance-rejection sampling
+              while (true) {
+                //sample x
+
+                double prob = (*dist)(*generator.getLocal());
+
+                Graph::Node nbr;
+                uint32_t p2;
+                FindSampleNeighbor(graph, curr, degree, prob, &nbr, &p2);
+
+                //sample y
+                double y = (*dist)(*generator.getLocal());
+                y = y * upper_bound;
+
+                //compute transition probability
+                double alpha;
+
+                //check if nbr is same as the previous node on this walk
+                if (nbr == prev) {
+                  alpha = prob_backward;
+                }  //check if nbr is also a neighbor of the previous node on this walk
+                else if (
+                    katana::FindEdgeSortedByDest(graph, prev, nbr) !=
+                    graph.edge_end(prev)) {
+                  alpha = 1.0f;
+                } else {
+                  alpha = prob_forward;
+                }
+
+                alpha = alpha * transition_matrix[p1][p2];
+                if (alpha >= y) {
+                  //accept y
+                  walk.push_back(std::move(nbr));
+                  types_vec.push_back(p2);
+                  break;
+                }
+              }  //end while
+
+            }  //end if-else loop
+          }    //end for
+
+          (*walks).push(std::move(walk));
+          (*types_walks).push(std::move(types_vec));
+        },
+        katana::steal(), katana::loopname("loop"));
+  }
+
+  //compute the histogram of edge types for each walk
+  katana::gstl::Vector<katana::gstl::Vector<uint32_t>>
+  ComputeNumEdgeTypeVectors(
+      const katana::InsertBag<katana::gstl::Vector<uint32_t>>& types_walks) {
+    katana::gstl::Vector<katana::gstl::Vector<uint32_t>> num_edge_types_walks;
+
+    katana::PerThreadStorage<
+        katana::gstl::Vector<katana::gstl::Vector<uint32_t>>>
+        per_thread_num_edge_types_walks;
+    katana::do_all(
+        katana::iterate(types_walks),
+        [&](const katana::gstl::Vector<uint32_t>& types_walk) {
+          katana::gstl::Vector<uint32_t> num_edge_types(numEdgeTypes + 1, 0);
+
+          for (auto type : types_walk) {
+            num_edge_types[type]++;
+          }
+
+          (*per_thread_num_edge_types_walks.getLocal())
+              .push_back(std::move(num_edge_types));
+        });
+
+    for (unsigned j = 0; j < katana::getActiveThreads(); ++j) {
+      for (auto num_edge_types :
+           *per_thread_num_edge_types_walks.getRemote(j)) {
+        num_edge_types_walks.push_back(std::move(num_edge_types));
```

Comment:
Ditto

This seems convoluted. The size of `num_edge_types_walks` is known. Instead of using `PerThreadStorage`, allocate `num_edge_types_walks` to the right size and then index into it to right to it - it should be a `do_all` loop.

Reply:
given that the types_walks is an InsertBag, I don't know how will get the desired index value

Reply:
Oh yes, that makes it more complicated. This is fine for now.